src/qibolab/compilers/compiler.py

```diff
@@ -141,27 +114,31 @@ def compile(self, circuit, platform):
         sequence = PulseSequence()
         # FIXME: This will not work with qubits that have string names
         # TODO: Implement a mapping between circuit qubit ids and platform ``Qubit``s
-        virtual_z_phases = defaultdict(int)
 
         measurement_map = {}
+        qubit_clock = defaultdict(int)
+        channel_clock = defaultdict(int)
         # process circuit gates
-        delays = defaultdict(int)
         for moment in circuit.queue.moments:
-            moment_start = sequence.finish
             for gate in set(filter(lambda x: x is not None, moment)):
                 if isinstance(gate, gates.Align):
                     for qubit in gate.qubits:
-                        delays[qubit] += gate.delay
+                        qubit_clock[qubit] += gate.delay
                     continue
-                gate_sequence, gate_phases = self._compile_gate(
-                    gate, platform, sequence, virtual_z_phases, moment_start, delays
-                )
-                for qubit in gate.qubits:
-                    delays[qubit] = 0
-
-                # update virtual Z phases
-                for qubit, phase in gate_phases.items():
-                    virtual_z_phases[qubit] += phase
+
+                rule = self[gate.__class__]
+                # get local sequence and phases for the current gate
+                gate_sequence = rule(gate, platform)
+                for pulse in gate_sequence:
+                    if qubit_clock[pulse.qubit] > channel_clock[pulse.qubit]:
+                        delay = qubit_clock[pulse.qubit] - channel_clock[pulse.channel]
+                        sequence.append(Delay(delay, pulse.channel))
+                        channel_clock[pulse.channel] += delay
+
+                    sequence.append(pulse)
+                    # update clocks
+                    qubit_clock[pulse.qubit] += pulse.duration
+                    channel_clock[pulse.channel] += pulse.duration
```

Comment:
This kind of follows the comment above. I don't want `Delay`s to have `qubit` attribute because this is not needed at the driver level. However, the compiler still needs to time things according to what qubit they are playing, so that for example a measurement (that will play on readout channel) will come after a gate (that will play on the drive channel) on the same qubit. This is what I am trying to achieve with `qubit_clock` and `channel_clock`.

For example
```
c = Circuit(1)
c.add(gates.X(0))
c.add(gates.M(0))
```
should be compiled to
```py
seq = PulseSequence()
seq.append(Pulse(..., PulseType.DRIVE, channel="drive0"))  # implements X
seq.append(Pulse(..., PulseType.READOUT, channel="readout0"))  # implements M
```